### ISSUE TITLE
Subscription to multiple entries in a double map does not reflect calls to remove_prefix

### ISSUE THREAD

**Issue by JoshOrndorff** [polkadot-js/api#1417]
When using `.multi` to to subscribe to multiple entries in a double map, the callback is always called, but not always with the correct value. When the change to storage was caused by a a call to `remove_prefix`, the callback contains the correct value (`None`) sometimes, but contains the previous value other times. The bug manifests itself more frequently when subscribed to more cells.

I originally discovered this when writing a [frontend for my tic tac toe game](https://github.com/JoshOrndorff/TicTacToe-frontend/blob/761296e1905a01d67c4fb57c517ee57d1c427bec/src/TicTacToe.jsx#L50-L54).

I then stripped that game down to a minimal reproducible example and wrote a script to demonstrate the problem. That work currently live on this [fairly obscure branch](https://github.com/JoshOrndorff/TicTacToe/tree/double-map-debug)

Essentially the script repeatedly sets and clears entries in a doublemap and queries the state in between each round. The output appears as follows. Notice the bug only shows up win the queries made using `.multi`

```
Alice is about to set 0,0
Cell 0,0 from single query: 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
Cell 0,0 from multi query : 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
Alice is about to clear 0
Cell 0,0 from single query:
Cell 0,0 from multi query :


Alice is about to set 0,0
Cell 0,0 from single query: 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
Cell 0,0 from multi query : 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
Alice is about to clear 0
Cell 0,0 from multi query : 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
Cell 0,0 from single query:


Alice is about to set 0,0
Cell 0,0 from single query: 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
Cell 0,0 from multi query : 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
Alice is about to clear 0
Cell 0,0 from single query:
Cell 0,0 from multi query :
```

In this small sample, the bug only showed up in 1 out of 3 queries. As mentioned above, the bug is easier to reproduce when subscribing to more cells. In fact I didn't even notice that it was intermittent when subscribed to a full game board of 9 cells.

I would be happy to help debug further, or clarify anything I've reported here.

**Comment by JoshOrndorff** [comment]
Good news, the race condition was in my test code, not the api.

When querying with a regular subscription as well as a multi (like I was doing thinking that would provide a control group) the bug is intermittent. When querying with _only_ a multi  subscription, the bug is perfectly reproducible.

Hopefully that helps debug.

**Comment by jacogr** [comment]
It does. It means it is related to the multi path (and value combination) instead of the linked logic.

**Comment by jacogr** [comment]
Ok, so I have not looked yet at the test-cases yet, but I believe I know exactly what is going wrong - and there is not way I can actually think of (atm), to get around this without severely hampering the dev experience elsewhere.

Assume we have 3 entries, A, B & C. So with the `subscribeStorage` does with multiple values is the following.

- initial subscribe, first result is the values for all, i.e. `[a0, b0, c0]`
- now assuming B changes, the result will be updated as `[null, b1, null]`
- assuming next A & C changes in the same block, the next result will be `[a2, null, c2]`
- assuming B now gets deleted, the next result is `[null, null, null]` 

Using the above is problematic from a dev perspective. Basically, you need to track the results on your own, i.e. when an empty value is received, you need to do magic. So the API does this for you and makes sure that you always  have the latest known values in all results, so the API will return for the above cases

- initial subscribe, `[a0, b0, c0]`
- B changes in block 1, `[a0, b1, c0]` (i.e. it returns the known values for A & C)
- A & C changes in  block 2, `[a2, b1, c2]` (fills in the known B)

This all goes swimmingly. Until B gets deleted. Now the result returned from the underlying is all `null`, however we have no way of knowing if -

- it has not actually updated
- it has been removed

So the API returns `[a2, b1, c2]`, the last known. 

So basically this is where we need to look. The "fill-in-the-gaps" behavior is something we don't want to remove, it really creates chaos for devs since they need to manage all the gaps in results themselves.

Will see what specifically happens in this case. But because we delete and the result is empty, i.e. `null` aka fill in the defaults from metadata, it is probably quite problematic since we cannot figure out what is the case of the `null`.

**Comment by JoshOrndorff** [comment]
That makes sense, and I'm glad to know it. It seems like the crux of the problem is

> we have no way of knowing if -
> -    it has not actually updated
> -    it has been removed

Are those optimized results (the ones that only contain the latest changes) coming from Substrate node RPC calls, or somewhere in the API. It seems necessary to distinguish those two cases.

**Comment by ianhe8x** [comment]
So is it that in the case that a value might be deleted, we should avoid using `.multi()`, and make subscription for each ?

**Comment by jacogr** [comment]
Closed in https://github.com/polkadot-js/api/pull/1732
